TTD/ReplayApi/TraceFallbacks/Fallbacks.cpp

```diff
@@ -0,0 +1,186 @@
+// even the tiniest bit of synchronization overhead can become very significant.
+static thread_local SegmentGatheredData s_segmentGatheredData{};
+
+// The data for multiple segments needs to be gathered into a single list for processing.
+// Each segment is replayed in a different thread, so a mutex is needed for synchronization.
+static std::mutex                       s_completedSegmentListMutex;
+static std::vector<SegmentGatheredData> s_completedSegmentList;
+
+// The thread continuity callback is invoked at the end of each segment, on the same thread that replayed the segment.
+// Its purpose is to get the data extracted by the high-frequency callbacks during replay of the segment,
+// compress, optimize or summarize it as appropriate, and add it to the global list.
+void __fastcall ThreadContinuityCallback(uintptr_t /*context*/) noexcept
+{
+    // The segment ended, so destructively remove and reset the TLS data.
+    // This is generally good practice because it avoids leaving potentially large
+    // data structures stuck in TLS for threads that might never use it again.
+    // We may wish to recycle objects or data buffers between segments,
+    // but in that case we should do that in a different way rather than leaving it
+    // in the TLS structure in case the thread replays a new segment.
+    SegmentGatheredData segmentGatheredData = std::exchange(s_segmentGatheredData, {});
+
+    {
+        std::lock_guard lock{ s_completedSegmentListMutex }; // Serialize access to global completed list.
+        s_completedSegmentList.push_back(std::move(segmentGatheredData));
+    }
+}
+
+// Gather raw fallback statistics from the replay engine, using the FallbackCallback.
+// For efficiency, we gather data per-segment in thread-local storage (TLS),
+// and then merge the per-segment data into a single map at the end.
+FallbackStatsMap GatherRawFallbackInfo(IReplayEngineView& replayEngineView)
+{
+    UniqueCursor pOwnedCursor{ replayEngineView.NewCursor() };
+    if (!pOwnedCursor)
+    {
+        throw std::runtime_error("Failed to create a replay engine cursor");
+    }
+
+    auto fallbackCallback = [&](bool synthetic, GuestAddress pc, size_t size, IThreadView const* pThreadView) {
+        // It's good practice to find the TLS data once.
+        // The compiler might not be able to optimize multiple accesses to the same TLS structure.
+        auto& segmentGatheredData = s_segmentGatheredData;
+
+        InstructionBytes fallback{};
+        DBG_ASSERT(size <= _countof(fallback.bytes));
+
+        fallback.size = static_cast<uint8_t>(size);
+        auto const queryResult = pThreadView->QueryMemoryBuffer(pc, BufferView{ fallback.bytes, size });
+
+        FallbackType const fallbackType = synthetic ? FallbackType::SyntheticInstruction : FallbackType::FullFallback;
+
+        Position const position = pThreadView->GetPosition();
+
+        InstructionBytes const& key = queryResult.Memory.Size == size ? fallback : InstructionBytes{};
```

Comment:
Fun thing: in the happy case, this makes a copy of `fallback` into a temporary (because both branches of the conditional need to be converted to a common type).

Prefer something like:

```suggestion
        if (queryResult.Memory.Size < size)
        {
            fallback.size = 0;
        }

        ... segmentGatheredData.Stats.insert({ fallback, value }) ...

```